### ISSUE TITLE
More efficient count/sum aggregation

### ISSUE THREAD

**Issue by splitice** [grafana/loki#6230]
**Is your feature request related to a problem? Please describe.**
Currently to get a count (or sum) of all events within a bucket the following format query has to be used:

```
sum by (key) (count_over_time({query} [1m]))
```

There are a few things to note here:
1. It would be faster without the counting over time
2. `[1m]` is required for the output value to be accurate, count by time always returns one bucket per minute.

**Describe the solution you'd like**

`sum by` and `count by` should be able to take a logql query directly and operate on those.

So what I would like to see is the above example become:
```
count by (key) ({query})
```

**Describe alternatives you've considered**

See above

**Comment by chaudum** [comment]
I think I need to clarify some things about metric queries here:

The `count_over_time()` is a range vector aggregation, which needs a time range over which it can aggregate logs; therefore the required `[1m]` range.

**Range (vector) aggregations** ([docs](https://grafana.com/docs/loki/latest/logql/metric_queries/#range-vector-aggregation)), such as `count_over_time` and `bytes_over_time` and [unwrapped range vector aggregations](https://grafana.com/docs/loki/latest/logql/metric_queries/#unwrapped-range-aggregations) are _the only way_ to "convert" log data into instant vectors.

The instant vector can then be used in **(instant) vector aggregations** ([docs](https://grafana.com/docs/loki/latest/logql/metric_queries/#built-in-aggregation-operators)), such as `sum`, `avg`, `min`, `max`, ... to aggregate values and group them by labels.

So your example query

```
sum by (key) (count_over_time({query} [1m]))
```

counts the amount of log lines that match the given query within the range of 1 minute (e.g. from `now-1m` to `now`) and returns that value for each matching stream at time `now`.
If you have e.g. 120 matching streams, you will have 120 values.

Then you group by a specific label (`key`) and aggregate the values by using the `sum` operator. This reduces the results to `n` values, where `n` is the amount of different values for `key`.

Simplifying the query into `sum/count by (key) ({query})` does not work, because `sum` only works on instant vectors, but the `query` returns a log lines. Also, without a range Loki would not know about over what time range it should aggregate the logs (count or bytes).
_Basically, you are missing important information how to extract values from the log data._ 

---

> **Note**:
> I talked about instant queries here, but same applies for range queries. They can be seen as a series of instant queries anyway.

**Comment by splitice** [comment]
If your issue is with the overloading of the `sum by` / `count by` syntax thats largely implementation specifics.

I personally don't see a problem with it as it aptly describes purpose (the aim of any language). However it could just as easily be an alternative syntax.

**Comment by chaudum** [comment]
> It sounds alot like your argument against the feature is that the current implementation doesnt do that. You do realize this is a feature issue not a bug right?

Yes, I understood that you suggested a "short form" of a query by combining two nested functions. However, with laying out how range aggregations and instant aggregation work, I wanted to show that your proposed short form lacks important information for achieve that the original query does.

> I'm well aware of what loki is currently doing. And the massive inefficiency of the workaround example. Hence the feature request.

I am not sure what inefficiency you mean? Do you mean that the query is too long to write?

> If your issue is with the overloading of the sum by / count by syntax thats largely implementation specifics.

Yes, overloading functions is indeed intentionally the avoided, because of their different usages.
PromQL also has a clear separation of **aggregation operations** ([docs](https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators)) and **\<aggregation\>_over_time** functions ([docs](https://prometheus.io/docs/prometheus/latest/querying/functions/#aggregation_over_time)).

> However it could just as easily be an alternative syntax.

An alternative/additional syntax of a short form would require to keep all the semantics and information the existing query has.

What you suggested:

```
sum by (key) (count_over_time({query} [1m])
=>
sum by (key) ({query})
```

would lose the information about the range over which it aggregates and which data (line count/bytes or label value) it aggregates, and is therefore ambiguous.

> Or are you suggesting that the correct syntax is (for a day):
> `sum by (key) (count_over_time({query} [1d]))`

I don't know exactly what you want to query, but yes, this query - executed as instant query - will return the amount of log lines grouped by `key` over the last day.
